Implement Notifications

Coder206 opened the pull request:
Implementation for notifications for the Janitor

Coder206 commented:
@jankeromnes It seems that the settings in my C9 environment are conflicting with yours as there are lots of lines with added spaces. Are you using a different setting than default when formatting?

Coder206 commented:
@jankeromnes The new logo in `png` is for Chrome support.

jankeromnes commented:
Thanks a lot for this pull request! I will have a look at your changes, and put any comments in the review.

> @jankeromnes It seems that the settings in my C9 environment are conflicting with yours as there are
> lots of lines with added spaces. Are you using a different setting than default when formatting?

Ah, are you making C9 auto-format your code files? I don't use that feature, so I'm not too familiar with its default settings (I thought they indented with 4 spaces by default, which I don't like). Janitor's current style has indeed a few more spaces here and there, which I believe make things more readable. Maybe there is a way to adjust C9's auto-format style parameters?

jankeromnes commented on static/js/janitor.js:
```diff
@@ -175,13 +175,57 @@ removeQueryString();
+// Settings: Notifications
+
+// TODO: 
+// Need to save the choice of the user (JSON? -- db.json notification: boolean)
```
I agree that the user's choice should be saved in `db.json`. Right now, the best way to achieve this is to make the checkbox trigger Ajax calls to the server, by implementing something similar to the following code:

/templates/settings-account.html:

```html
          <form id="notifications-form" class="col-sm-12 col-md-8 account-form ajax-form has-feedback">
            <label class="control-label">Enable notifications</label>
            <input class="form-control" data-submit-on="change" name="enabled" type="checkbox">
          </form>
```

/static/js/settings.js:

```js
ajaxForm('#notifications-form', 'notifications', function (form, data) {
  var status = 'error';
  var message = data.message;
  if (data.status === 'enabled' || data.status === 'disabled') {
    status = 'success';
  }
  updateFormStatus(form, status, message);
});
```

/app.js:

```js
  app.ajax.on('notifications', (data, end, query) => {
    let { user } = query.req;
    if (!user || !('enabled' in data)) {
      end();
      return;
    }

    let { enabled } = data;
    user.settings.notifications = enabled;
    db.save();

    end({ status: (enabled ? 'enabled' : 'disabled') });
  });
});
```

jankeromnes replied:
Additionally, for `user.settings.notifications` to actually exist, you would need to change `/lib/users.js` like so:

```diff
    user = {
      email: email,
      keys: {
        cloud9: '',
        cloud9user: '',
        ssh: {
          fingerprint: '',
          privateKey: '',
          publicKey: ''
        }
      },
+     settings: {
+       notifications: false
+     },
      machines: {},
      data: {}
    };
```

Then we'd also need to upgrade all `db.json` files out there to the new format (fortunately, there is only one of these in production today, but eventually we'll have to come up with a way to automatically upgrade DB formats during updates, e.g. on server startup right after an update).

jankeromnes commented on static/js/janitor.js:
```diff
@@ -175,13 +175,57 @@ removeQueryString();
+
+// TODO: 
+// Need to save the choice of the user (JSON? -- db.json notification: boolean)
+// Welcome notification 
```
It's a good idea to trigger one notification when the user enables them, maybe something like "You will now receive notifications like this when there is a change in your contributions."?

jankeromnes commented on static/js/janitor.js:
```diff
@@ -175,13 +175,57 @@ removeQueryString();
+// C9 events notifications handler
+// Janitor events handler
+// Fallback (onscreen) for older browsers
+// Interactive UI/UX for notifications (ie: Settings{1})
```
See my suggestion above on how to make it work via Ajax. I think at some point we'll need to improve the Settings panel with categories and better UX.

jankeromnes commented on static/js/janitor.js:
```diff
@@ -175,13 +175,57 @@ removeQueryString();
+  }
+  else {
+    alert("Unchecked")
+      // Confirm that there will be no notifications?
```
If you use the suggested Ajax code, the checkbox and "Enable notifications" label will turn green when an Ajax call was successful, or red if there was an error. This is quite subtle though, so maybe at some point we'll want to confirm things in more obvious ways, e.g. a small in-page notification (like Android's "toast" notifications maybe?)

jankeromnes commented on static/js/janitor.js:
```diff
@@ -175,13 +175,57 @@ removeQueryString();
+  }
+}
+
+function spawnNotification(theTitle, theBody, theIcon = "/img/janitor.png") {
```
Nit: No need to prefix parameter names with "the" here. Also, the front-end code is not yet ES6, so default values for parameters might not be supported. Eventually, I agree that we should also upgrade the front-end code to ES6.

jankeromnes commented on static/js/janitor.js:
```diff
@@ -175,13 +175,57 @@ removeQueryString();
+      body: theBody,
+      icon: theIcon
+  }
+  var n = new Notification("The Janitor: " + theTitle, options);
```
Nit: No need to create the `n` variable, as it's never used. (Additionally, in the local style we would prefer more explicit variable names, e.g. `notification`).

jankeromnes commented on static/js/janitor.js:
```diff
@@ -175,13 +175,57 @@ removeQueryString();
+      icon: theIcon
+  }
+  var n = new Notification("The Janitor: " + theTitle, options);
+}
```
Nit: Please keep a `\n` character at the end of the file, otherwise Git and GitHub will complain.

jankeromnes commented on static/js/settings.js:
```diff
@@ -32,3 +32,10 @@ ajaxForm('#cloud9-form', 'key', function (form, data) {
+// Settings: Notifications checkbox event listener
+
+document.getElementById("notify").addEventListener("click", function(){
+  toggleNotifications();
```
Nit: In general, I think it's more robust to use the checkbox's actual state, rather than just toggling an internal state when there is a click (for example, if there is a race condition somewhere, we might end up with a checkbox that does the opposite of what it's supposed to do). Incidentally, the Ajax call I suggested uses the checkbox's actual state, so you could reuse that to enable/disable notifications.

jankeromnes commented on static/js/settings.js:
```diff
@@ -32,3 +32,10 @@ ajaxForm('#cloud9-form', 'key', function (form, data) {
+
+document.getElementById("notify").addEventListener("click", function(){
+  toggleNotifications();
+});
```
Nit: Please add a trailing `\n`.

jankeromnes commented on templates/settings-account.html:
```diff
@@ -1,24 +1,31 @@
+      <form id="cloud9user-form" class="col-sm-6 col-md-4 account-form ajax-form has-feedback">
+        <input type="hidden" name="name" value="cloud9user">
+        <label class="control-label">Cloud9 Username</label>
+        <input class="form-control" data-submit-on="blur" name="key" placeholder="Username" type="text" value={{=u ser.keys.cloud9user in json}}>
```
It seems you accidentally changed the pre-existing `{{= user` code to `{{=u ser`, which will probably fail.

jankeromnes commented on templates/settings-account.html:
```diff
@@ -1,24 +1,31 @@
-            <textarea class="form-control" data-submit-on="blur" name="key" placeholder="SSH key" rows="7">{{= user.keys.cloud9 in html}}</textarea>
-            <p>Please add your <a target="_blank" href="https://c9.io/account/ssh" title="Your public SSH key on c9.io">Cloud9 SSH key</a> to enable Cloud9 IDE access.</p>
-          </form>
-        </div>
```
Nit: Please don't modify the indentation of this file. I agree that it seems silly, but this template file is actually a fragment of a complete web page, and once all fragments are joined (with `header.html` and `footer.html`) to serve a full Janitor web page, you'll see in "view source" that everything is indented properly. Additionally, not re-indenting this file will make this pull request easier to review, because GitHub will be able to highlight exactly which characters you actually changed, and which code you left unchanged.

Coder206 commented:
@jankeromnes Thanks for the suggestions! I appreciate! I hope to find some time to implement them.

notriddle commented:
Protip: add `?w=1` to the end of a URL with a diff on it to get it to ignore all the whitespace-only changes.

jankeromnes commented:
Thank you @notriddle! I'm very fond of `git show -w` or `git diff -w`, and I always forget that it's also possible on GitHub.

Coder206 commented:
@jankeromnes I fixed all the nits that you asked and I brought in some of the code you recommended into the PR. Over the next few days I hope to make more progress. Thanks! :)

jankeromnes commented on static/js/settings.js:
```diff
@@ -32,3 +32,16 @@ ajaxForm('#cloud9-form', 'key', function (form, data) {
+  var message = data.message;
+  if (data.status === 'enabled' || data.status === 'disabled') {
+    status = 'success';
+    // TODO: Welcome message
```
👍

jankeromnes commented on static/js/settings.js:
```diff
@@ -32,3 +32,16 @@ ajaxForm('#cloud9-form', 'key', function (form, data) {
+
+// Settings: Enable Notifications 
+
+ajaxForm('#notifications-form', 'notifications', function (form, data) {
```
Nit: We're in the process of making the coding style a little denser than before. No need to fixup the other functions in this file, but please only use one empty line above the function comment, and remove the empty line between the comment and the function.

jankeromnes commented on static/js/settings.js:
```diff
@@ -32,3 +32,16 @@ ajaxForm('#cloud9-form', 'key', function (form, data) {
+    // TODO: Welcome message
+  }
+  updateFormStatus(form, status, message);
+});
```
Nit: Please leave a `\n` at the end of the file.

jankeromnes commented on templates/settings-account.html:
```diff
@@ -20,5 +20,12 @@ <h2 class="header">Settings</h2>
           </form>
         </div>
+        <div class="row">
+          <!-- TODO: position the checkbox -->
```
To position the checkbox properly, I suggest adding it to the same "row" as the "Email" and "Cloud9 username" fields, by changing their size classes to `col-sm-4 col-md-3` (you can use the same for your `<form id="notifications-form"`, so that the whole "row" takes up "12" units when the page is "small", and "9" when the page is "medium").

Actually if you do that, please also update the `<form id="cloud9-form"` from `col-md-8` to `col-md-9` as well, so that it vertically aligns properly on the right on "medium" screens.

Additionally, [bootstrap's docs](http://getbootstrap.com/css/#forms) suggest the following markup for checkboxes:

```html
<div class="checkbox">
  <label>
    <input type="checkbox"> Check me out
  </label>
</div>
```

so I would suggest remodeling your form like so (no `class="form-control"` on the checkbox please, that's what makes it look "gigantic"):

```html
<form id="notifications-form" class="col-sm-4 col-md-3 account-form ajax-form has-feedback">
  <label class="control-label">Notifications</label>
  <div class="checkbox">
    <label>
      <input data-submit-on="change" name="enabled" type="checkbox">
      Enable notifications
    </label>
  </div>
</form>
```

Then, once you've implemented the `/$notifications` Ajax call in `app.js`, you'll see the checkbox and labels turn green or red after clicking it, depending on whether the Ajax call succeeded or not.

Coder206 commented:
@jankeromnes I made the changes requested and I also updated the style of the comments above functions for this file. My next step is to write the backend in the next few days.

Coder206 commented on static/js/settings.js:
```diff
@@ -1,9 +1,7 @@
 
-
 // Settings: Cloud9 username form.
-
```
It was on my way while scrolling.

jankeromnes replied:
Additionally, feel free to add your name to the copyright notice in files you modify :) (I also update the year whenever I make big changes in a file).

jankeromnes commented on templates/settings-account.html:
```diff
@@ -11,9 +11,18 @@ <h2 class="header">Settings</h2>
             <label class="control-label">Cloud9 Username</label>
             <input class="form-control" data-submit-on="blur" name="key" placeholder="Username" type="text" value={{= user.keys.cloud9user in json}}>
           </form>
+          <form id="notifications-form" class="col-sm-4 col-md-3 account-form ajax-form has-feedback">
```
Nit: If you want the top row to be properly split into three equal columns, the "Email" and "Cloud9 username" forms will also need to be updated to the `col-sm-4 col-md-3` class.

Coder206 commented:
@jankeromnes I hope all is well! I was wondering if you know how to rectify this:

![page-shot-2017-3-13 janitor_notify - cloud9](https://cloud.githubusercontent.com/assets/7758155/23875407/b6745d90-080f-11e7-8d4d-e8c60c67fd99.png)

 I tried using `git fetch upstream master` with no success. Please there was a syntax issue with my app.js file, so I replaced it with the current via copy/paste before performing the fetch.

jankeromnes commented:
Hi @Coder206 ! First of all, I'm so sorry that you're running into so many issues with the Janitor code base. These are really unnecessary traps or pit falls that take too much time and energy to work around. The good news is that I'm making great progress on putting Janitor in Janitor, and in fact it will be my next pull request. Soon this should allow you to completely forget about these problems, so you can spend more time on what you actually planned to do!

Now about the actual issue you're seeing, they happen because by default, `app.js` asks `lib/boot.js` to generate new HTTPS certificates. However, when you're developing locally (or in Cloud9), you don't usually expose the HTTP port, so Let's Encrypt can't verify that you actually own your domain (by checking a secret token that should be served on your HTTP port).

Soon, there will be a `db.get('security').forceHttp` parameter allowing you to bypass this generation (my next PR), but in the meantime, you can just disable HTTPS certificate generation altogether by commenting out this line from `app.js`'s required setup tasks:

```js
 boot.executeInParallel([
   boot.forwardHttp,
   // boot.ensureHttpsCertificates,
   boot.ensureDockerTlsCertificates
 ], () => {
```

jankeromnes commented on api/user-api.js:
```diff
@@ -217,3 +217,65 @@ configurationAPI.put({
+  }]
+});
+
+notificationsAPI.isEnabled({
```
Ah yes, Self API does some magic, but doesn't support things like this. Here you are building a "REST API", which exposes some features that can be triggered by calling specific URLs (e.g. `/api/user/notifications`) with specific HTTP verbs (e.g. `GET`, `POST`, `DELETE`, etc) and other parameters (e.g. query parameters like `?thing=value` in the URL, a request body containing JSON, etc).

I guess what would make sense for a notifications API would be the following:
- `GET /api/user/notifications` could return the list of all previous notifications
- `POST /api/user/notifications` could be used to create/push a new notification (and in the request body, you could send a JSON object that contains all the information required to produce a new notification)

Also, regarding how to change the value of the boolean `enabled`, you can do that pretty much any way. I would suggest the following API:

- `GET /api/user/notifications/enabled` could return a boolean (whether notifications are enabled or not)
- `PUT /api/user/notifications/enabled` could set `user.notifications.enabled` to the value sent via a specific parameter (e.g. `PUT /api/user/notifications/enabled?enabled=true`; or `PUT /api/user/notifications/enabled` with a request body like `{ "enabled": true }`)

Such an API could be implemented like this:

```js
const notificationsEnabledAPI = notificationsAPI.api('/enabled');

notificationsEnabledAPI.get({
  handler: (request, response) => {
    // handler that simply returns the value of `user.notifications.enabled`, e.g.
    response.json({ enabled: user.notifications.enabled }, null, 2);
  }
});

notificationsEnabledAPI.put({
  handler: (request, response) => {
    // handler that accepts an `enabled` parameter; sets `user.notifications.enabled` to it; and maybe does a few other things to properly enable or disable notifications, e.g.
    const enabled = request.query.enabled === 'true'; // this comes from a URL query parameter like "/api/user/notifications/enabled?enabled=true"
    user.notifications.enabled = enabled;
    db.save();
    if (enabled) {
      /* ... */
    } else {
      /* ... */
    }
  }
});
```

jankeromnes commented on api/user-api.js:
```diff
@@ -217,3 +217,65 @@ configurationAPI.put({
+  }]
+});
+
+notificationsAPI.enable({
```
This should probably be `notificationsEnabledAPI.put({})` (see previous comment).

jankeromnes commented on api/user-api.js:
```diff
@@ -217,3 +217,65 @@ configurationAPI.put({
+  }
+});
+
+notificationsAPI.disable({
```
You could handle both `enable` and `disable` actions in the same `notificationsEnabledAPI.put({})` handler (see previous comment).

jankeromnes commented on lib/routes.js:
```diff
@@ -207,12 +207,29 @@ exports.contributionsPage = function (response, user = null) {
+const notificationsTemplate = camp.template('./templates/notifications.html');
+exports.notificationsPage = function (response, user) {
+  const title = 'My Notifications';
+
```
You could do this in the HTML template by iterating over `user.notifications.feed` (or whatever property name you'd like to give to the list of all notifications).

jankeromnes commented on lib/users.js:
```diff
@@ -233,6 +233,9 @@ function getOrCreateUser (email) {
       machines: {},
+      notifications: {
+        enabled: false,
+      },
```
I agree! You could add an Array property like `feed: []` or `list: []` or `history: []` or similar.

jankeromnes commented on templates/header.html:
```diff
@@ -27,6 +27,9 @@
           <ul class="nav navbar-nav navbar-right">{{if user then {{
             <li><a href="/projects/">Projects</a></li>
             <li><a href="/contributions/">Contributions</a></li>
+            <!-- TODO: Only show when notifications are enabled -->
```
You can do this by writing a `{{if user.notifications.enabled then {{`, similar to the `if user` above.

jankeromnes commented on templates/header.html:
```diff
@@ -27,6 +27,9 @@
             <li><a href="/contributions/">Contributions</a></li>
+            <!-- TODO: Only show when notifications are enabled -->
+            <!-- TODO: Number of unread notifications using the API GET notifications -->
+            <li><a href="/notifications/">Notifications  <span class="badge badge-red">42</span></a></li>
```
👍

jankeromnes commented on lib/users.js:
```diff
@@ -273,6 +276,13 @@ function getOrCreateUser (email) {
+    user.notifications = {
+      enabled: false,
+    };
+  }
```
👍

jankeromnes commented on lib/routes.js:
```diff
@@ -207,12 +207,29 @@ exports.contributionsPage = function (response, user = null) {
   account: camp.template('./templates/settings-account.html'),
   configurations: camp.template('./templates/settings-configurations.html'),
   integrations: camp.template('./templates/settings-integrations.html'),
+  notifications: camp.template('./templates/settings-notifications.html'),
```
👍

jankeromnes commented on api/user-api.js:
```diff
@@ -217,3 +217,65 @@ configurationAPI.put({
   }]
 });
+
+const notificationsAPI = userAPI.api('/notifications');
```
I guess this list of notifications could probably be a Bootstrap [List group](http://getbootstrap.com/components/#list-group), with different styling for read/unread state.

As an example, you could do something similar to the [list of integrations](https://github.com/JanitorTechnology/janitor/blob/d00be427b25ed1ab3afd86934b163834f1c67c5a/templates/settings-integrations.html#L11-L20) template.

If you need advice on how to get a bootstrap markup that looks the way you want, I'd be happy to help 😄 Feel free to draw something up (e.g. on paper, or Paint, or whatever you find most convenient) and I can tell you how to achieve it.

jankeromnes commented on templates/settings-header.html:
```diff
@@ -5,6 +5,7 @@
             <li role="presentation" {{if {{section === "account"}} then {{class="active"}}}}><a href="/settings/account/">Account</a></li>
             <li role="presentation" {{if {{section === "configurations"}} then {{class="active"}}}}><a href="/settings/configurations/">Configurations</a></li>
             <li role="presentation" {{if {{section === "integrations"}} then {{class="active"}}}}><a href="/settings/integrations/">Integrations</a></li>
+            <li role="presentation" {{if {{section === "notifications"}} then {{class="active"}}}}><a href="/settings/notifications">Notifications</a>
```
👍

jankeromnes commented on api/user-api.js:
```diff
@@ -217,3 +217,65 @@ configurationAPI.put({
+      response.json({ error: 'Unauthorized' });
+      return;
+    }
+    response.json(user.notifications, null, 2);
```
Sure, the way to do that with our current template system is like so:

```html
  <div class="checkbox">
    <label>
      <input{{if user.notifications.enabled then {{ checked="true"}}}} data-submit-on="change" name="enabled" type="checkbox"> Enabled
    </label>
  </div>
```

jankeromnes commented on templates/settings-notifications.html:
```diff
@@ -0,0 +1,15 @@
+          <div class="col-sm-12">
+            <p>Configure your preferences for receiving notifications related to your code.</p>
+          </div>
+          <form action="/api/user/notifications" class="account-form ajax-form has-feedback col-sm-6" data-refresh-after-success="true" method="isEnabled">
```
Nit: If you implement my `PUT /api/user/notifications/enabled` suggestion above, the `method` needs to be `"put"`.

Nit: Also, you probably don't want to refresh the page after a successful request, so you could remove the `data-refresh-after-success` attribute.

jankeromnes commented on templates/notifications.html:
```diff
@@ -0,0 +1,5 @@
+    <section class="notifications">
+      <div class="container">
+        <h2 class="header">My Notifications</h2>
+      </div>
```
I guess this list of notifications could probably be a Bootstrap [List group](http://getbootstrap.com/components/#list-group), with different styling for read/unread state.

As an example, you could do something similar to the [list of integrations](https://github.com/JanitorTechnology/janitor/blob/d00be427b25ed1ab3afd86934b163834f1c67c5a/templates/settings-integrations.html#L11-L20) template.

If you need advice on how to get a bootstrap markup that looks the way you want, I'd be happy to help 😄 Feel free to draw something up (e.g. on paper, or Paint, or whatever you find most convenient) and I can tell you how to achieve it.

jankeromnes commented on templates/settings-notifications.html:
```diff
@@ -0,0 +1,15 @@
+            <label>
+              <!-- TODO: Toggle checkbox based on notifications status -->
+              <input data-submit-on="change" name="enabled" type="checkbox">Enable notifications
+            </label>
```
Sure, the way to do that with our current template system is like so:

```html
  <div class="checkbox">
    <label>
      <input{{if user.notifications.enabled then {{ checked="true"}}}} data-submit-on="change" name="enabled" type="checkbox"> Enabled
    </label>
  </div>
```

jankeromnes replied:
One more thing: I believe that our current `<form>` submission code doesn't support checkboxes, because their value comes from `input.checked` instead of `input.value`. You might need to make a small tweak to [/js/janitor.js](https://github.com/JanitorTechnology/janitor/blob/d00be427b25ed1ab3afd86934b163834f1c67c5a/static/js/janitor.js#L223) to check if `element.type === "checkbox"`, and in that case use `element.checked` instead of `element.value`.

Coder206 commented on api/user-api.js:
```diff
@@ -217,3 +217,75 @@ configurationAPI.put({
+    }
+    else {
+      response.statusCode = 500; // Internal Server Error
+      response.json({ error: 'Could not change the state of notifications for this user' }, null, 2);
```
Would this be required or could we simply end with [line 282](https://github.com/JanitorTechnology/janitor/pull/39/files#diff-dec71030eda346e491be4e56b83344e2R282)?

jankeromnes replied:
I think this code makes it look like an error happened on the server when you simply disable notifications (i.e. call `PUT /api/user/notifications/enabled` with `enabled = false`).

Coder206 commented on api/user-api.js:
```diff
@@ -217,3 +217,75 @@ configurationAPI.put({
+      response.json({ error: 'Unauthorized' });
+      return;
+    }
+    //TODO: Build the notification from the request and send it to the db
```
I am not certain how to go about this. 
Does this make sense:
`{ 'notifications.feed': { icon: { src: 'request.icon.path', alt: 'request.icon.tag', }, title: 'request.title', body: 'request.body', action: 'request.action'} }, null, 2)`

There is also a field 'read' and I think it is a necessity to keep this off the API (as it is managed internally), is there a preferred and more secure way to protect it from tampering from the API than asking only for the fields above?

jankeromnes replied:
I think you could parse the request body as JSON (see example [here](https://github.com/JanitorTechnology/janitor/blob/d00be427b25ed1ab3afd86934b163834f1c67c5a/api/hosts-api.js#L188-L200)) and then extract the notification details from it.

Also, in general when there are properties you don't want to be part of the API (not visible in GET or modifying in POST etc) you can make the API work on a sub-property of an item, and store non-API properties in another sub-property, e.g.

```js
// GET:
response.json(user.notifications.feed.map({ notification, read } => notification), null, 2);

// POST:
const notification = JSON.parse(json);
user.notifications.feed.push({ notification, read: false });
```

This way the API never mentions any `read` property even though we use it internally, because the API only uses the `user.notifications[i].notification` sub-property (we do similar things for `user.profile`, `host.properties`, etc. which are modified by the API even though the containing objects are not).

Coder206 commented on lib/routes.js:
```diff
@@ -207,12 +207,31 @@ exports.contributionsPage = function (response, user = null) {
+const notificationsTemplate = camp.template('./templates/notifications.html');
+exports.notificationsPage = function (response, user) {
+  const title = 'My Notifications';
+  const notifications = db.get('user.notifications.feed');
```
Would this suffice for iterating over the feed (I inspired myself from the projects template)?

jankeromnes replied:
Nit: Actually, `db.get('user.notifications.feed')` would try to get a property like this in `db.json`:

```js
{
  "user.notifications.feed": "some value"
}
```

What you actually want is to use `user.notifications.feed` directly in the template (no need to create a special `notifications` template variable here, since you can find it withing the `user` in the template directly).

Coder206 commented on static/css/janitor.css:
```diff
@@ -361,6 +361,11 @@ h4 {
+/* Notifications Tab */
+.unread {
+  border-bottom: solid #E36200;
+}
```
I am planning on showing an orange bar below the notifications that are unread similar to the settings tabs like Configurations:

![page-shot-2017-6-26 settings the janitor](https://user-images.githubusercontent.com/7758155/27566919-4e8e1942-5ab6-11e7-94ad-a50649d2621e.png)

jankeromnes replied:
Interesting suggestion!

I believe Twitter's unread notifications simply have a light blue background that becomes white when you have read the notification.

Coder206 commented on static/js/janitor.js:
```diff
@@ -220,7 +220,11 @@ function getFormData (form) {
+        data[name] = element.checked;
+      } else {
+        data[name] = element.value;
+      }
```
I have not been able to test this functionality as of yet.

jankeromnes replied:
Looks correct! You could also make it shorter like this:

```js
data[name] = element.type === "checkbox" ? element.checked : element.value;
```

Coder206 commented on templates/header.html:
```diff
@@ -27,6 +27,8 @@
             <li><a href="/projects/">Projects</a></li>
             <li><a href="/contributions/">Contributions</a></li>
+            <!-- TODO: Number of unread notifications using the API GET notifications -->
+            {{if users.notifications.enabled then {{ <li><a href="/notifications/">Notifications  <span class="badge badge-red">{user.notifications.unread}</span></a></li> }}}}
```
This functionality was the reason why I think the [unread](https://github.com/JanitorTechnology/janitor/pull/39/files#diff-dec71030eda346e491be4e56b83344e2R282) field is useful, it would make for less computation on the user's side at this state. Increasing performance and ultimately reducing the time between the API receiving a notification sharing it with the affected user.

jankeromnes replied:
Actually templates are computed server-side before sending them to users, but I believe that the cost of such a `user.notifications.feed.filter(n => !n.read).length` computation is negligible for now.

Coder206 commented on templates/notifications.html:
```diff
@@ -0,0 +1,22 @@
+          <div class="col-sm-12">
+            <ul class="list-group">
+              <!-- TODO: Iterate over the template below for each notification within  --> 
+              <!-- START - Template for notifications -->
```
I would like to dynamically create this as new notifications are shown. Would it be possible to create them [here](https://github.com/JanitorTechnology/janitor/pull/39/files#diff-52503f2920dcbf5082edfc5670ede0f1R217) Perhaps we could implement a similar system as Twitter's infinite scroll where the notifications visible to the user are the only that are rendered (the rest are hidden). If this is the route we decided, I think performance is important, I find it awkward to wait to a blank screen while a whole new set of notifications/posts appear.

Coder206 commented on templates/settings-notifications.html:
```diff
@@ -0,0 +1,14 @@
+          </form>
+        </div>
+      </div>
+    </section>
```
TODO: Add a new line here.

Coder206 commented on lib/users.js:
```diff
@@ -233,6 +233,22 @@ function getOrCreateUser (email) {
       machines: {},
+      notifications: {
+        enabled: false,
+        unread: 1,
```
To find out the number of notifications displayed, it think it's best that we count the number as we add notifications and subtract this field's values as they become read as opposed to looping over the whole array in field and checking the read status.

jankeromnes replied:
This is an interesting idea, but by not using the actual states withing the `feed` parameter, we expose ourselves to bugs where this `unread` shorthand variable becomes out-of-sync with the array.

Then again, I agree that when we'll have very long feeds, it might be inefficient to iterate over the whole array.

In these cases, I generally try to start with the less complex / simpler solution (in this case, I would prefer leaving the `unread` optimization for later, and simply call `user.notifications.feed.filter(n => !n.read).length` when needed), but I'll leave this choice up to you :)

Coder206 commented on lib/users.js:
```diff
@@ -233,6 +233,22 @@ function getOrCreateUser (email) {
+        unread: 1,
+        feed: [
+          {
+            icon: {
```
To allow for custom images to be displayed as notifications. The required values are taken by the  [notifications tab](https://github.com/JanitorTechnology/janitor/pull/39/files#diff-c5b0e3321db6c30500da15f992b83212R10)

Coder206 commented on lib/users.js:
```diff
@@ -233,6 +233,22 @@ function getOrCreateUser (email) {
+            title: 'Welcome to the Janitor! :)',
+            body: 'Please join us on IRC for support.',
+            action: 'https://kiwiirc.com/client/irc.freenode.net/?#janitor',
+            read: false,
```
This determines whether to display the unread CSS style (the bottom or left border)

Coder206 commented on api/user-api.js:
```diff
@@ -217,3 +217,75 @@ configurationAPI.put({
+      return;
+    }
+    response.json({ enabled: user.notifications.enabled }, null, 2);
+  }
```
TODO: Add an example

Coder206 commented on api/user-api.js:
```diff
@@ -217,3 +217,75 @@ configurationAPI.put({
+    else {
+      response.statusCode = 500; // Internal Server Error
+      response.json({ error: 'Could not change the state of notifications for this user' }, null, 2);
+    }
```
TODO: Add an example
TODO: Add a newline

jankeromnes replied:
E.g.

```js
examples: [{
  request: {
    queryParameters: { enabled: true }
  },
  response: {
    body: JSON.stringify({ enabled: true }, null, 2)
  }
}]
```

jankeromnes commented on api/user-api.js:
```diff
@@ -217,3 +217,75 @@ configurationAPI.put({
+  }]
+});
+
+notificationsAPI.put({
```
Maybe using the POST method would make more sense for sending a notification than the PUT method? Not sure, just a random suggestion :)

jankeromnes commented on api/user-api.js:
```diff
@@ -217,3 +217,75 @@ configurationAPI.put({
+  },
+  examples: [{
+    response: {
+      body: JSON.stringify({ 'notifications.feed': { icon: { src: 'Path to the image displayed with the notification', alt: 'Image tag', }, title: 'Notification title', body: 'Description of the notification', action: 'Link to bring the user to act on the notification'} }, null, 2)
```
For requests that create things (e.g. a notification to be sent), you can either return the created object (here it would be the notification object) or simply set the status to `201 // Created` and call `response.end()`.

jankeromnes commented on lib/routes.js:
```diff
@@ -207,12 +207,31 @@ exports.contributionsPage = function (response, user = null) {
+  const notifications = db.get('user.notifications.feed');
+
+  response.template({
+    notifications,
```
Nit: You can remove this variable from the template parameters.

jankeromnes commented on lib/users.js:
```diff
@@ -273,6 +289,26 @@ function getOrCreateUser (email) {
+        {
+          icon: {
+            src: '/img/janitor.svg',
+            alt: 'Janitor Logo',
```
How important is the `alt` attribute for notification icons? Just curious.

Coder206 replied:
I think this could be beneficial for accessibility. Despite this, I noticed the [settings-integrations](https://github.com/JanitorTechnology/janitor/blob/master/templates/settings-integrations.html#L12) has it for it's images, so I followed it's lead.

jankeromnes commented on lib/users.js:
```diff
@@ -273,6 +289,26 @@ function getOrCreateUser (email) {
+          },
+          title: 'Welcome to the Janitor! :)',
+          body: 'Please join us on IRC for support.',
+          action: 'https://kiwiirc.com/client/irc.freenode.net/?#janitor',
```
👍

Coder206 commented on api/user-api.js:
```diff
@@ -217,3 +217,108 @@ configurationAPI.put({
+  // Example with the first notification
+  examples: [{
+    request: {
+      body: JSON.stringify({ icon: '/img/janitor.svg', title: 'Welcome to the Janitor! :)', body: 'Please join us on IRC for support.', action: 'https://kiwiirc.com/client/irc.freenode.net/?#janitor' }, null, 2)
```
I think using the notification users will first receive will be beneficial in that they can visualize the outcome.

Coder206 commented on api/user-api.js:
```diff
@@ -217,3 +217,108 @@ configurationAPI.put({
+    response.json(user.notifications.feed.map(({ notification }) => notification), null, 2);
+  },
+  examples: [{
+    response: {
```
An example of two notifications being returned by the API.

Coder206 commented on api/user-api.js:
```diff
@@ -217,3 +217,108 @@ configurationAPI.put({
+    user.notifications.enabled = enabled;
+    response.json({ enabled: user.notifications.enabled }, null, 2);
+    console.log(user.notifications.enabled);
+    console.log(enabled);
```
These `logs` are to help in debugging an issue with the checkbox. In this commit, they always return `false` even when the checkbox is checked and unchecked.

jankeromnes replied:
I believe this is because the front-end code responsible for calling this API handler doesn't actually set a query parameter `?enabled=true` but sends a JSON request body that looks like `{"enabled": true}`.

Coder206 commented on lib/users.js:
```diff
@@ -273,6 +286,33 @@ function getOrCreateUser (email) {
+          read: false,
+        },
+        {
+          notification: {
```
Just a simple `test` notification, to be deleted in the near future.

Coder206 commented on static/css/janitor.css:
```diff
@@ -361,6 +361,15 @@ h4 {
+}
+
+.notification {
+  margin-bottom: 1px;
```
It provides a space to avoid cutting the border-bottom (if we choose to take that route, see this [Discourse](https://discourse.janitor.technology/t/implementing-notifications/40/6) conversation post.

Coder206 commented on templates/notifications.html:
```diff
@@ -0,0 +1,20 @@
+        <div class="row">
+          <div class="col-sm-12">
+            <ul class="list-group">{{ for (const { notification, read } of user.notifications.feed) { }}
+              <li {{if (!read) then {{= class='list-group-item notification unread' in html}} else {{= class='list-group-item notification' in html}}}}>
```
Would there be more preferred way to write the `if`? Is it possible to write this line similar to line 6's `for` loop?

jankeromnes replied:
You could simplify this template macro like so:

```html
<li class="list-group-item notification {{if !read then {{unread}}}}">
```

Coder206 commented on lib/users.js:
```diff
@@ -273,6 +286,33 @@ function getOrCreateUser (email) {
+  // Temporary migration code: Previous users didn't have notification settings.
+  //if (!user.notifications) {
+    user.notifications = {
+      enabled: true,
```
Please note this is set to `true` while [line 313](https://github.com/JanitorTechnology/janitor/pull/39/files#diff-dec71030eda346e491be4e56b83344e2R313) of `user-api.js` returns `false`

Coder206 commented on templates/settings-notifications.html:
```diff
@@ -0,0 +1,14 @@
+          </div>
+          <form action="/api/user/notifications/enabled" class="account-form ajax-form has-feedback col-sm-6" method="put">
+            <label>
+              <input {{if user.notifications.enabled then {{ checked="true"; value="false" }} else {{ checked="false"; value="true" }}}} data-submit-on="change" name="enabled" type="checkbox">Enable notifications
```
I tried modifying the value for the `input` so that the API can be send the new status of `notifications.enabled`. However, from [line 313](https://github.com/JanitorTechnology/janitor/pull/39/files#diff-dec71030eda346e491be4e56b83344e2R313) of `user-api.js` it appears that this form is not able to change the status for the user. If I refresh, the value stays the same (likely due to the fact the browser kept in memory the choice) but when I switch to another page and return to the notifications settings it returns to the state I set at [line 293](https://github.com/JanitorTechnology/janitor/pull/39/files#diff-93aa3cd6516e4575bf93e79539665635R293) of `user.js`

Coder206 replied:
I would also like to change the visibility of the notification tab in the header (the page to receive notifications) to display only when the checkbox is enabled. With the current code it works, but it requires to reload the server and hard code the values at [line 293](https://github.com/JanitorTechnology/janitor/pull/39/files#diff-93aa3cd6516e4575bf93e79539665635R293) of `user.js`.

jankeromnes replied:
Nit: Please simplify this checkbox like so:

```html
<input {{if user.notifications.enabled then {{checked="checked"}}}} data-submit-on="change" name="enabled" type="checkbox"> Enable notifications
```

jankeromnes replied:
Ah right, to show the "Notifications" menu entry you would need to refresh the page. I'll try to find a good suggestion for this when I'll test this feature in the near future.

Coder206 commented:
I think the problems with the checkbox is that the data is only submitted when the box is checked, according to: https://stackoverflow.com/questions/11424037/does-input-type-checkbox-only-post-data-if-its-checked 

@jankeromnes I will try to work in this considering the above so feel free to ignore my comments in the last review

jankeromnes commented on api/user-api.js:
```diff
@@ -217,3 +217,108 @@ configurationAPI.put({
+      response.json({ error: 'Unauthorized' });
+      return;
+    }
+    const enabled = request.query.enabled === 'true';
```
Nit: Please verify that the API call actually includes an `enabled` query parameter (e.g. if it is not included, `request.query.enabled` will be `undefined`, and since `undefined !== 'true'`, your API handler will believe it should disable notifications, whereas no `enabled` parameter was actually specified).

You can verify it like so:

```js
if ('enabled' in request.query) {
  user.notifications.enabled = (request.query.enabled === 'true');
  response.json(...);
  return;
}

let json = '';
request.on('data' // etc
request.on('end', () => {
  let body = JSON.parse(json); // with a try/catch
  if ('enabled' in body) {
    user.notifications.enabled = body.enabled;
  }
```

(note: I've also added a suggestion to support JSON request bodies that look like `{ "enabled": true }`)

jankeromnes commented on api/user-api.js:
```diff
@@ -217,3 +217,108 @@ configurationAPI.put({
+    console.log(user.notifications.enabled);
+    console.log(enabled);
+  },
+  examples: [{
```
Nit: Please add an empty line before `examples`.

jankeromnes commented on api/user-api.js:
```diff
@@ -217,3 +217,108 @@ configurationAPI.put({
+  },
+  examples: [{
+    request: {
+      queryParameters: { enabled: true }
```
👍 

If you end up supporting both query parameters and JSON request body, please add a second example with:

```js
request: {
  body: JSON.stringify({ enabled: true }, null, 2)
}
```

jankeromnes commented on templates/header.html:
```diff
@@ -26,7 +26,9 @@
             <li><a href="/projects/">Projects</a></li>
-            <li><a href="/contributions/">Contributions</a></li>
+            <li><a href="/contributions/">Contributions</a></li>{{if user.notifications.enabled then {{
+            <li><a href="/notifications/">Notifications {{ let unread = user.notifications.feed.filter(n => !n.read).length; if (unread > 0) { }}
```
Nit: Sorry for writing such horrible code :P 

Please improve its readability slightly by adding a line break between `{{` and `let unread = ...`.

Nit: Please use full variable names (i.e. `notification` instead of just `n`).

jankeromnes commented on templates/notifications.html:
```diff
@@ -0,0 +1,20 @@
+          <div class="col-sm-12">
+            <ul class="list-group">{{ for (const { notification, read } of user.notifications.feed) { }}
+              <li {{if (!read) then {{= class='list-group-item notification unread' in html}} else {{= class='list-group-item notification' in html}}}}>
+                <img class='logo-small' src={{= notification.icon in json}}>
```
Nit: Please use double-quotes in HTML code (i.e. `"` instead of `'`).

jankeromnes commented on templates/settings-notifications.html:
```diff
@@ -0,0 +1,14 @@
+      <div class="container">
+        <div class="row row-header">
+          <div class="col-sm-12">
+            <p>Configure your preferences for receiving notifications related to your code.</p>
```
Nit: I find "to your code" to be somewhat ambiguous and misleading. Maybe it's not necessary to mention it?

Coder206 commented on api/user-api.js:
```diff
@@ -217,3 +217,159 @@ configurationAPI.put({
+      json += String(chunk);
+    });
+    
+    request.on('end', () => {
```
Within the Firefox Dev Tools, I can change the state of `enabled`. Unfortunately, I was receiving 400 errors when making a request for posting a notification, it appears `request.on('end'...)` is never triggered after receiving the JSON object that would resemble this: 
```

  {
    "icon": "/img/janitor.svg",
    "title": "Welcome to the Janitor! :)",
    "body": "Please join us on IRC for support.",
    "action": "https://kiwiirc.com/client/irc.freenode.net/?#janitor"
  },
```

Despite this, I tried it with the same code for [configurations](https://github.com/JanitorTechnology/janitor/blob/master/api/user-api.js#L51-L69) and experienced the same symptoms as before.

I also tried to individually add the fields, eg:

1st request:

```
{
    "icon": "/img/janitor.svg",
}
```

2nd request:

```
{
    "title": "Hello world!",
}
```

This resulted in a JSON array similar to the following:

```
[
  {
    "icon": "/img/janitor.svg",
    "title": "Welcome to the Janitor! :)",
    "body": "Please join us on IRC for support.",
    "action": "https://kiwiirc.com/client/irc.freenode.net/?#janitor"
  },
  {
    "icon": "/img/janitor.svg",
    "title": "Your Container Firefox-1 is ready :)",
    "body": "Happy coding!",
    "action": "https://c9.io"
  },
  { 
    "icon": "/img/janitor.svg"
  },
  { 
    "title": "Hello world!"
  },
]
```

Note the separation of the different fields to different notifications. That makes sense because they were sent in different requests.

This leads me to believe that I made a mistake with the API request for POSTing the notification object. The problem is that I am unsuccessful in properly making this API POST.

Coder206 replied:
So the correct way to post to the notifications API is the following: 
```

[
  {
    "icon": "/img/janitor.svg",
    "title": "Welcome to the Janitor! :)",
    "body": "Please join us on IRC for support.",
    "action": "https://kiwiirc.com/client/irc.freenode.net/?#janitor"
  }
]
```

Coder206 replied:
The correct way to POST a notifications is having a Request body like so: 

```
{
    "icon": "/img/janitor.svg",
    "title": "Welcome to the Janitor! :)",
    "body": "Please join us on IRC for support.",
    "action": "https://kiwiirc.com/client/irc.freenode.net/?#janitor"
 }
```

Coder206 commented on templates/notifications.html:
```diff
@@ -0,0 +1,20 @@
+        <h2 class="header">My Notifications</h2>
+        <div class="row">
+          <div class="col-sm-12">
+            <ul class="list-group">{{ for (const {notification, read} of user.notifications.feed.slice().reverse()) { }}
```
This helps by making a copy of the JSON feed.notification object and reverses it to show the latest notification first.

jankeromnes replied:
Interesting idea! Note: You could also simply use `notifications.feed.unshift()` instead of `.push()` in order to keep the notifications array permanently reversed.

Coder206 commented on lib/users.js:
```diff
@@ -273,6 +286,33 @@ function getOrCreateUser (email) {
   });
+  
+  // Temporary migration code: Previous users didn't have notification settings.
+  if (!user.notifications) {
```
When I visited this `if` statement after we chatted live, it was commented out, should we uncomment it like so?

Coder206 replied:
In the future commit I will remove the `if` statement from this code.

jankeromnes commented on static/js/settings.js:
```diff
@@ -32,3 +32,16 @@ ajaxForm('#cloud9-form', 'key', function (form, data) {
+// Settings - Notifications tab: Change the state of notifications
+
+var form = document.getElementById('notifications-form');
+var checkbox = document.getElementById('notifications-state');
```
Nit: You can get this checkbox with `form.elements['enabled']` (using the input name), no need for an extra `id` attribute.

jankeromnes commented on static/js/settings.js:
```diff
@@ -32,3 +32,16 @@ ajaxForm('#cloud9-form', 'key', function (form, data) {
+    form.setAttribute('action', '/api/user/notifications/enabled?enabled=true');
+  } else {
+    form.setAttribute('action', '/api/user/notifications/enabled?enabled=false');
+  }
```
Nit: Please simplify this code by using a ternary operator like this:

```js
form.setAttribute('action', '/api/user/notifications/enabled?enabled=' + (checkbox.checked ? 'true' : 'false'));
```

jankeromnes commented on templates/settings-notifications.html:
```diff
@@ -0,0 +1,14 @@
+          </div>
+          <form id="notifications-form" action="/api/user/notifications/enabled?enabled={{= !user.notifications.enabled}}" data-class="account-form ajax-form has-feedback col-sm-6" data-refresh-after-success="true" method="put">
+            <div class="col-sm-12">
+              <input id="notifications-state" autocomplete="off" {{if user.notifications.enabled then {{ checked="checked" }}}} data-submit-on="change" name="enabled" type="checkbox"><label class="notifications-label" for="checkbox">Enable notifications</label>
```
Nit: Please remove the unnecessary `id` attribute (see prior nit with `form.elements['enabled']`), and please place this `input` into a `label` element like in bootstrap's examples (this allows clicking on the text to change the checkbox's state, which is convenient).

Coder206 commented on templates/notifications.html:
```diff
@@ -0,0 +1,21 @@
+        <div class="row">
+          <div class="col-sm-12">
+            <!-- TODO: Change the state of notifications from read to unread on toggle and click -->
+            <ul class="list-group">{{ for (const {notification, read} of user.notifications.feed.unshift()) { }}
```
I have success with `user.notifications.feed.slice().reverse()` however when using `user.notifications.feed.unshift()`, the Notifications page does not load (similar to what happens when a mistake in the template is made)

jankeromnes replied:
Yes, that's because `Array.unshift(item)` should be used when pushing an item into an array (it's similar to `Array.push(item)`, but adds it at position `0` instead of adding it at the end. This would give you a `user.notifications.feed` array that's already "reversed" by design).

However, I think you should ignore my suggestion to use `unshift`, and keep using `feed.reverse()`, because you may need stable notifications IDs, whereas `unshift` will change all notification IDs every time you push an item.

TL;DR - my bad!

Coder206 commented on templates/header.html:
```diff
@@ -5,7 +5,12 @@
+    {{if user.notifications.enabled then {{ 
+    {{ let unread = user.notifications.feed.filter(notification => !notification.read).length; if (unread > 0) { }}
+    {{= {{ "(" + unread + ")" }} in html}} {{ } }} }}}}
+    {{= title in html}}{{if title then {{ • }}}}The Janitor 
```
Adds the number of notifications that are unread to a tab. 

TODO: Work on making this dynamic (so user's will receive updates live).

Coder206 commented on static/css/janitor.css:
```diff
@@ -361,6 +361,18 @@ h4 {
+.unread {
+  border-left: solid #E36200;
+}
+
```
With the current design of notifications, the previous margin adjustment becomes obsolete:

![screenshot-2017-6-28 my notifications the janitor 4](https://user-images.githubusercontent.com/7758155/27757012-b21ce01c-5dca-11e7-86cb-6dc6a3e6ca38.png)

The previous code helped to avoid a cut-off in the bottom border of this style of notification unread state:

![screenshot-2017-6-28 my notifications the janitor](https://user-images.githubusercontent.com/7758155/27757031-e731b016-5dca-11e7-9192-e2a7b5a93536.png)

Coder206 replied:
@jankeromnes Using Firefox's DevTools, it says for the all notifications `border-radius: 4px` [abridged from the actual CSS codes (ie `border-**bottom**/**top**-**right**/**left**-radius: 4px)

Coder206 commented on lib/users.js:
```diff
@@ -233,6 +232,20 @@ function getOrCreateUser (email) {
+            title: 'Welcome to the Janitor! :)',
+            body: 'Please join us on IRC for support.',
+            action: 'https://kiwiirc.com/client/irc.freenode.net/?#janitor',
+          },
```
To be removed

Coder206 commented on lib/routes.js:
```diff
@@ -207,12 +207,29 @@ exports.contributionsPage = function (response, user = null) {
+  response.template({
+    title,
+    user,
+    scripts: []
```
The console logs have been saying `checkbox is null` in the settings tabs excluding the notifications (where checkbox is found). To rectify this, I would to make a `settings-notifications.js` file to separate the code for notifications from the other settings tabs.

Coder206 commented on lib/users.js:
```diff
@@ -273,6 +286,33 @@ function getOrCreateUser (email) {
+    user.notifications = {
+      enabled: false,
+      feed: [
+        {
```
> notifications.isValid(notificationObjectParsedInTheAPI) 

Excellent idea! I am planning on postponing the work on the verification for notifications POSTs until this feature is implemented as I envision it to be the verification for all the APIs related to notifications. This module could potentially verify all API requests from the notifications API. 

On a similar note, I am still debating the best method to change the state of a notification from unread to read and vice-versa. I was thinking of assigning IDs to the notifications and modifying their state with an API. Despite this, I prefer (for simplicity in `db.json`) to find a way to simply have a click event on the notification and it can modify the state of the specific notification while avoiding IDs. I think the previous could be another addition to the new module. Let me know what you think.

Coder206 commented:
@jankeromnes Thanks for pointing out all the nits in my code, I am happy to perfect the appearance any legibility! :) I have added some code for implementing page title notifications as well that are useful if a user is away from the website. I would like to improve this by alerting about unread notifications without the need for a user's interaction. Do you know of way this could be done without an page refresh? Perhaps there could be a watcher to look for new notification periodically?

jankeromnes commented:
Thanks a lot for your great work! I'll do another review pass and try to help you where I can. 😄

> Adds the number of notifications that are unread to a tab.
> 
> TODO: Work on making this dynamic (so user's will receive updates live).

>  I have added some code for implementing page title notifications as well that are useful if a user is away from the website. I would like to improve this by alerting about unread notifications without the need for a user's interaction. Do you know of way this could be done without an page refresh? Perhaps there could be a watcher to look for new notification periodically?

Great idea with the title! Maybe there is also a way to make the browser highlight the tab when something happens? (I think some websites are highlighted with a special color or round icon when there are updates).

Also, in order to make these changes dynamic in the page, I'll need to add Server-Sent Events support to the website and API. This is something I was planning to do for some time (to enable live updates on the website, for things like container status, live stats, etc) but I'll have to do it sooner if you plan on using it for notifications.

You can either implement live updates (title, tab state, red badge, etc) in a follow-up pull request, or wait for my SSE implementation to land before finishing this pull request. 😃

jankeromnes commented on templates/notifications.html:
```diff
@@ -0,0 +1,21 @@
+                </li>
+              </a>
+              {{ } }}
+            </ul>
```
Nit: Please put the `{{ } }}` right next to `</a>` in order to produce cleaner HTML code (in its current form, this template adds a new line with lots of trailing spaces after each `<a></a>` block).

jankeromnes commented on templates/settings-notifications.html:
```diff
@@ -0,0 +1,16 @@
+          <div class="col-sm-12">
+            <p>Configure your preferences for receiving notifications.</p>
+          </div>
+          <form id="notifications-form" action="/api/user/notifications/enabled?enabled={{= !user.notifications.enabled}}" data-class="account-form ajax-form has-feedback col-sm-6" data-refresh-after-success="true" method="put">
```
Nit: It would be great to escape `{{= !user.notifications.enabled}}` with a [fleau parser](https://github.com/espadrine/fleau#parsers). Please either use the `in uri` parser, or implement a more precise one at [the top](https://github.com/JanitorTechnology/janitor/blob/135c06adc3a64fabba84c21eac9b62c76b24a2ec/lib/routes.js#L15-L23) of the `/lib/routes.js` file, e.g.:

```js
camp.templateReader.parsers.boolean = text => {
  return String(!!text);
};
```

(This returns the string `"false"` if `text` is "falsy" (i.e. `false`, `0`, `null`, `undefined`, `""`, etc.) and `"true"` if `text` contains something sensible. A notable caveat of this parser is that `boolean("false")` will produce `"true"` (because the string "false" is a valid string that's considered "truthy" because it does contain "something sensible"). However, it would be the implementor's fault for calling `in boolean` on something that's not boolean in the first place (it's the same if you do something like `if ("false") { /* this code will run! */ }`)).

You could even contribute it to the upstream [fleau repository](https://github.com/espadrine/fleau) if you feel like it! (But still implement it in `/lib/routes.js` so that we can merge this pull request before a new fleau is released).

jankeromnes commented on templates/settings-notifications.html:
```diff
@@ -0,0 +1,16 @@
+          <form id="notifications-form" action="/api/user/notifications/enabled?enabled={{= !user.notifications.enabled}}" data-class="account-form ajax-form has-feedback col-sm-6" data-refresh-after-success="true" method="put">
+            <div class="col-sm-12">
+              <label>
+                <input id="notifications-state" autocomplete="off" {{if user.notifications.enabled then {{ checked="checked" }}}} data-submit-on="change" name="enabled" type="checkbox"><span class="notifications-label">Enable notifications</span>
```
Nit: Since this `input` has a `name` attribute, and lives inside a `form` that we can query, there is no need for an `id` attribute.

Coder206 commented:
@jankeromnes I have addressed the nits you mentioned above. The API allows exclusively for `booleans` in the requests for notifications settings of the user. I also omitted the POST notification, the badge and the page title notifications in this PR but will be added in a follow-up PR as discussed. 

So this PR adds:

- The notifications page (hidden from the user for now)
- The settings section for notifications
- The logic for setting the state of notification settings of the user

I would great appreciate some feedback. My apologies for the delays in pushing this PR. 

Also, is my build ok? (Travis)

jankeromnes commented on lib/users.js:
```diff
@@ -33,7 +33,7 @@ exports.get = function (request, callback) {
       return;
     }
-
+    
```
Nit: You're adding some trailing spaces here. Please remove them, it's cleaner without them. 😄

jankeromnes commented on static/js/settings.js:
```diff
@@ -32,3 +32,17 @@ ajaxForm('#cloud9-form', 'key', function (form, data) {
+
+  checkbox.addEventListener('change', function () {
+   form.setAttribute('action', '/api/user/notifications/enabled?enabled='
+     + (checkbox.checked ? true : false));
```
Nit: These `true` and `false` booleans will get coerced into Strings anyway (by the `+` in front of them), so you might as well use Strings directly, as in `(checkbox.checked ? 'true' : 'false')`.

jankeromnes commented on templates/settings-notifications.html:
```diff
@@ -0,0 +1,16 @@
+          <div class="col-sm-12">
+            <p>Configure your preferences for receiving notifications.</p>
+          </div>
+          <form id="notifications-form" action="/api/user/notifications/enabled?enabled={{= !user.notifications.enabled in boolean}}" data-class="account-form ajax-form has-feedback col-sm-6" data-refresh-after-success="true" method="put">
```
Nit: Maybe it's silly, but we try to order most things in alphabetical order, like for example HTML element attributes. Here, `id` comes after `data-refresh-after-success` but before `method`.

jankeromnes commented on api/user-api.js:
```diff
@@ -217,3 +217,119 @@ configurationAPI.put({
+  
+  examples: [{
+    request: {
+      queryParameters: { enabled: true }
```
Hm, this doesn't seem to appear in the documentation page (`/reference/api`). Maybe it's because it should be a string instead of a boolean, i.e. `'true'` instead of `true`? Or this is a Self API bug. I need to investigate.

jankeromnes replied:
I just had a look, and it seems that `request.queryParameters` are only used [for testing](https://github.com/JanitorTechnology/selfapi/blob/103e12c311e21aa7249a5662cae8e861b0375d70/selfapi.js#L259-L267), and not for [HTML](https://github.com/JanitorTechnology/selfapi/blob/103e12c311e21aa7249a5662cae8e861b0375d70/selfapi.js#L406) or [Markdown](https://github.com/JanitorTechnology/selfapi/blob/103e12c311e21aa7249a5662cae8e861b0375d70/selfapi.js#L473) documentation exports. This is because the documentation only shows the "canonical" path (i.e. `/api/user/notifications/enabled`) whereas `queryParameters` can be set differently for every example. Maybe we could repeat the path and add query parameters for examples that have them? (We don't currently repeat the API handler path in request examples, only specific headers or a body if those are present).

The good news is, your example will work in testing (we're about to enable tests for the whole project in #102 (hooray!)), but they won't show up in documentation examples for now, which could make the documentation somewhat confusing (as in: "How do I specify that `enabled` should be `"true"` or `"false"`? It doesn't show in the documentation").

In other news, maybe you can add an additional example, i.e.

```js
request: {
  body: JSON.stringify({ enabled: true }, null, 2)
}
```

And if you make this the first example, the documentation will automatically be fixed (since it's based on the first example, and here the first example would no longer use query parameters).

Coder206 replied:
Would it be also possible to have multiple `request`s in the examples? (`body: /api/user/notifications/enabled?enabled="true"` works as you mentioned)

Coder206 commented on templates/header.html:
```diff
@@ -5,7 +5,7 @@
     <meta name="theme-color" content="#333">
     <meta name="viewport" content="width=device-width, initial-scale=1">
-    <title> {{= title in html}}{{if title then {{ • }}}}The Janitor </title>
+    <title>{{= title in html}}{{if title then {{ • }}}}The Janitor</title>
```
To be reverted

jankeromnes commented:
Thanks for another round of commits! I'll have a look.

Also, apologies for missing your questions earlier, with the many comments already in this thread I only discovered it now, so I'll reply:

> I have addressed the nits you mentioned above. The API allows exclusively for booleans in the requests for notifications settings of the user. I also omitted the POST notification, the badge and the page title notifications in this PR but will be added in a follow-up PR as discussed.

That's great! Sounds good to me. 👍

> So this PR adds:
> 
> The notifications page (hidden from the user for now)

That's great for "testing in production" 😄 I did the same for the current project stats pages, e.g. https://janitor.technology/projects/servo/ (they're still not linked to from anywhere on the website, but eventually it will be part of a sort of "project dashboard").

> The settings section for notifications
> The logic for setting the state of notification settings of the user
> I would great appreciate some feedback. My apologies for the delays in pushing this PR.

No worries at all, this project already benefits from a lot from the spare time you generously invest in it! However, apologies for the delays in reviewing it.

> Also, is my build ok? (Travis)

Please disregard any failures, I simply enabled Travis before actually enabling or fixing all the tests. Since #102 was merged recently, future pushes based on it should start showing more useful Travis feedback. 😄

jankeromnes commented on templates/settings-notifications.html:
```diff
@@ -0,0 +1,16 @@
+          <div class="col-sm-12">
+            <p>Configure your preferences for receiving notifications.</p>
+          </div>
+          <form action="/api/user/notifications/enabled?enabled={{= !user.notifications.enabled in boolean}}" data-class="account-form ajax-form has-feedback col-sm-6" data-refresh-after-success="true" id="notifications-form" method="put">
```
Nit: I believe by `data-class` you actually meant `class`. 😉

jankeromnes commented on templates/settings-notifications.html:
```diff
@@ -0,0 +1,16 @@
+            <p>Configure your preferences for receiving notifications.</p>
+          </div>
+          <form action="/api/user/notifications/enabled?enabled={{= !user.notifications.enabled in boolean}}" data-class="account-form ajax-form has-feedback col-sm-6" data-refresh-after-success="true" id="notifications-form" method="put">
+            <div class="col-sm-12">
```
Nit: Since your `form` already has `col-sm-6`, no need to specify a column class here. However, it seems that Bootstrap checkboxes should be inside a `<div class="checkbox">`, as shown in [this example](http://getbootstrap.com/css/#forms-example).

So please replace `col-sm-12` here with `checkbox` (it will already set the right spacing between the box and the label text, and it will use a "pointer" cursor on the label text, making it clearer that it's clickable).

jankeromnes commented on templates/settings-notifications.html:
```diff
@@ -0,0 +1,16 @@
+          <form action="/api/user/notifications/enabled?enabled={{= !user.notifications.enabled in boolean}}" data-class="account-form ajax-form has-feedback col-sm-6" data-refresh-after-success="true" id="notifications-form" method="put">
+            <div class="col-sm-12">
+              <label>
+                <input autocomplete="off" {{if user.notifications.enabled then {{ checked="checked" }}}} data-submit-on="change" name="enabled" type="checkbox"><span class="notifications-label">Enable notifications</span>
```
Nit: If you use the `class="checkbox"` indicated above, you can get rid of the `class="notifications-label"` here, and actually of the entire `span`. Also, please put the label text on a new line, to make the template more readable.

jankeromnes commented on static/css/janitor.css:
```diff
@@ -361,6 +365,12 @@ h4 {
+
+.notifications-label {
+  padding-left: 10px;
+}
```
Nit: If you use a `<div class="checkbox">` as indicated below, you'll no longer need this class. In that case, please remove it.

jankeromnes commented on api/user-api.js:
```diff
@@ -217,3 +217,122 @@ configurationAPI.put({
+          break;
+        default:
+          response.statusCode = 400; // Bad Request
+          response.json({ error: 'Parameter \'enabled\' should be "true" or "false"' }, null, 2);
```
Oops, my bad, it seems that ScoutCamp actually parses `?enabled=true` query parameters into actual booleans. Please simplify this code to: 

```js
if ('enabled' in request.query) {
  const { enabled } = request.query;
  if (typeof(enabled) !== "boolean") {
    response.statusCode = 400; // Bad Request
    response.json({ error: 'Parameter \'enabled\' should be "true" or "false"' }, null, 2);
    return;
  }
  user.notifications.enabled = enabled;
  db.save();
  response.json({ enabled }, null, 2);
  return;
}
```

Coder206 commented on lib/users.js:
```diff
@@ -33,7 +33,7 @@ exports.get = function (request, callback) {
       callback(null);
       return;
     }
-
```
No sure what is happening over here. Shouldn't both of the modifications cancel each other out? ( `-` line and `+` back the previous line)

jankeromnes replied:
What's happening is that you accidentally added trailing spaces on this line, so Git understands this as "remove the line with no spaces" and "add a new line with 4 spaces".

Coder206 commented:
@jankeromnes I am not sure why Travis is complaining. I am experiencing the same output locally.

jankeromnes commented:
Thanks for another round! I'll take a look.

> I have addressed the latest set of nits and the code is working nominally.

Indeed! Let's see if it goes through max Q without RUD. 😄

> Also, what are your thoughts on renaming the commit message for 1f6e786 to something for meaningful like "Implementing part of notifications API and UX for interacting with notifications."

As you prefer. I think I would use a message like `Add an API and a Settings panel to enable/disable user notifications.`, but your message might be clearer. 🙂

> I am not sure why Travis is complaining. I am experiencing the same output locally.

So the message `Error: Can't set headers after they are sent.` means that you are calling `response.json()` and/or `response.end()` twice on the same `response` object. I'll have a look at your code to see if I can spot anything suspicious.

jankeromnes commented on api/user-api.js:
```diff
@@ -217,3 +217,123 @@ configurationAPI.put({
+  }]
+});
+
+// API sub-resource to manage a user's notification settings.
```
Nit: Maybe `API sub-resource to enable or disable user notifications.` might be clearer.

jankeromnes commented on api/user-api.js:
```diff
@@ -217,3 +217,123 @@ configurationAPI.put({
+  
+  examples: [{
+    request: {
+      body: JSON.stringify({ enabled: true }, null, 2),
```
Nit: You can also simply use `'{ enabled: true }'` as request body (JSON requests don't have to be prettified, only our API's JSON responses should always use `JSON.stringify(message, null, 2)`).

jankeromnes commented on api/user-api.js:
```diff
@@ -217,3 +217,123 @@ configurationAPI.put({
+      }
+    });
+    
+    response.json({ enabled: user.notifications.enabled }, null, 2);
```
It looks like you're no longer calling `response.json()` multiple times in a row, as in the previous patch. Maybe the Travis build is out-of-date (as demonstrated by the `api/user-api.js:345:18` in its stack trace, which is a line that no longer exist in your current revision).

jankeromnes commented on static/js/settings.js:
```diff
@@ -32,3 +32,17 @@ ajaxForm('#cloud9-form', 'key', function (form, data) {
+  checkbox.addEventListener('change', function () {
+   form.setAttribute('action', '/api/user/notifications/enabled?enabled='
+     + (checkbox.checked ? "true" : "false"));
+   }, false);
```
Nit: Actually, since you're using `data-refresh-after-success`, you don't need to update the form's action anymore. Please undo the changes to this file.

jankeromnes commented on templates/notifications.html:
```diff
@@ -0,0 +1,19 @@
+              <a href={{= notification.action in json}}>
+                <li class="list-group-item notification {{if !read then {{list-group-item-stressed}}}}">
+                  <img class="logo-small" src={{= notification.icon in json}}>
+                    <div class="list-group-item-details">
```
Nit: Here the indentation makes it look like your `div` is nested in the `img` element above, which isn't possible because `img` is a "void element" (can't have any content). Please de-indent this `div` by 2 spaces.

jankeromnes commented:
- [x] code review passed
- [x] linting passed
- [x] all tests passed

Let's merge this! 😄

jankeromnes commented:
Thanks a lot for implementing the foundation of our notification system @Coder206 ! And congratulations on landing your first big change (second after fef7a00cffef8b92c55c9b20be5743dc5cbdaa32) in this code base! Huge 👍 to you 🎉🎆💯
